crates/ra_mbe/src/lib.rs

```diff
@@ -337,4 +337,40 @@ SOURCE_FILE@[0; 40)
         );
     }
 
+    #[test]
+    fn test_two_ident() {
+        let rules = create_rules(
+            r#"
+        macro_rules! foo {
+            ($ i:ident, $ j:ident) => (
+                fn $ i() { $j }
+            )
+        }        
+"#,
+        );
+        assert_expansion(&rules, "foo! { bar, woo }", "fn bar () {woo}");
+    }
+
+    /// The following tests are port from intellij-rust directly
+    /// https://github.com/intellij-rust/intellij-rust/blob/c4e9feee4ad46e7953b1948c112533360b6087bb/src/test/kotlin/org/rust/lang/core/macros/RsMacroExpansionTest.kt
+
+    #[test]
+    fn test_path() {
+        let rules = create_rules(
+            r#"
+        macro_rules! foo {
+            ($ i:path) => {
+                fn foo() { let a = $ i; }
+            }
+        }
+"#,
+        );
+        assert_expansion(&rules, "foo! { foo }", "fn foo () {let a = foo ;}");
+        assert_expansion(
+            &rules,
+            "foo! { bar::<u8>::baz::<u8> }",
+            "fn foo () {let a = bar :: < u8 > :: baz :: < u8 > ;}",
```

Comment:
I wonder if implementing a minimal formatter for rule-based spacing around operators (like this: https://github.com/intellij-rust/intellij-rust/blob/c41f1855baf7e5929552f6198c39afa52bc1cd68/src/main/kotlin/org/rust/ide/formatter/impl/spacing.kt#L46-L52) would make the tests more readable?

Reply:
Sounds good to me, i will create another PR for that later.